src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,242 @@
+                var argumentNullException = compilation.GetTypeByMetadataName(ArgumentNullExceptionName);
+                if (argumentNullException is null)
+                {
+                    return;
+                }
+
+                var argumentNullExceptionConstructor = argumentNullException.InstanceConstructors.FirstOrDefault(m =>
+                    m.DeclaredAccessibility == Accessibility.Public
+                    && m.Parameters.Length == 1
+                    && m.Parameters[0].Type.SpecialType == SpecialType.System_String);
+                if (argumentNullExceptionConstructor is null)
+                {
+                    return;
+                }
+
+                var objectType = compilation.GetSpecialType(SpecialType.System_Object);
+                var referenceEqualsMethod = objectType
+                    .GetMembers(nameof(ReferenceEquals))
+                    .OfType<IMethodSymbol>()
+                    .FirstOrDefault(m => m.DeclaredAccessibility == Accessibility.Public && m.Parameters.Length == 2);
+                // This seems sufficiently rare that it's not a big deal to just not perform analysis in this case.
+                if (referenceEqualsMethod is null)
+                {
+                    return;
+                }
+
+                context.RegisterSyntaxNodeAction(context => AnalyzeSyntax(context, argumentNullExceptionConstructor, referenceEqualsMethod), SyntaxKind.MethodDeclaration, SyntaxKind.LocalFunctionStatement);
+            });
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, IMethodSymbol argumentNullExceptionConstructor, IMethodSymbol referenceEqualsMethod)
+        {
+            var cancellationToken = context.CancellationToken;
+
+            var semanticModel = context.SemanticModel;
+            var syntaxTree = semanticModel.SyntaxTree;
+
+            var option = context.Options.GetOption(CodeStyleOptions2.PreferParameterNullChecking, semanticModel.Language, syntaxTree, cancellationToken);
+            if (!option.Value)
+            {
+                return;
+            }
+
+            var node = context.Node;
+            var block = node switch
+            {
+                MethodDeclarationSyntax methodDecl => methodDecl.Body,
+                LocalFunctionStatementSyntax localFunctionStatement => localFunctionStatement.Body,
+                _ => throw ExceptionUtilities.UnexpectedValue(node)
+            };
+            if (block is null)
+            {
+                return;
+            }
+
+            var methodSymbol = semanticModel.GetDeclaredSymbol(node, cancellationToken);
+            if (methodSymbol is null)
+            {
+                return;
+            }
+
+            foreach (var statement in block.Statements)
+            {
+                switch (statement)
+                {
+                    // if (param == null) { throw new ArgumentNullException(nameof(param)); }
```

Comment:
hrmm... i think you need to go further.  except for strings, i think you need to check that this is not an overloaded operator right?  because `!!` wont' be legal.  def test cases liek that, as well as interesting cases of ref-types, value-types, nullable-types, typeparameter-types (constrained and otherwise).

Reply:
As a heuristic, I would consider saying that if the `==` is defined on a special type, then treat it as equivalent to the null check we perform via `!!`. For example: string is a special type, so we don't mind that a non-`object ==` operator is used.

Reply:
does this work for string though?  we likely would want this feature to light up, and strings don't have a built-in (unless i'm mistaken).

Reply:
It does actually do the thing we want on string (string.operator == appears to be considered built in)